google/cloud/status_or.h

```diff
@@ -24,33 +24,52 @@ namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 /**
- * `StatusOr<T>` is used to return a value or an error status.
+ * `StatusOr<T>` represents either a usable `T` value or a `Status` object
+ * explaining why a `T` value is not present. Typical usage of `StatusOr<T>`
+ * looks like usage of a smart pointer, or even a std::optional<T>, in that you
+ * first check its validity using a conversion to bool (or by calling
+ * `StatusOr<T>::ok()`), then you may dereference the object to access the
+ * contained value. It is undefined behavior (UB) to dereference a
+ * `StatusOr<T>` that is not "ok". For example:
  *
- * If the library is compiled with exceptions disabled there is no standard C++
- * mechanism to report errors. In that case we use this class to wrap the values
- * returned to the application.
+ * @code
+ * StatusOr<Foo> foo = FetchFoo();
+ * if (!foo) {  // Same as `if (!foo.ok())`
+ *   // handle error and probably look at foo.status()
+ * }
+ * foo->DoSomethingFooey();
```

Comment:
Do you want to highlight that using `foo` is UB if that `if()` branch falls through?

```
  * foo->DoSomethingFooey(); // UB if !foo
```

Reply:
Good point. done.